README.md
# [base64](https://waitingsong.github.io/base64/)

Base64 encoding/decoding in pure JS on both modern Browsers and Node.js based on [base64-js](https://www.npmjs.com/package/base64-js).  
Also supports [URL-safe base64](https://en.wikipedia.org/wiki/Base64#URL_applications)

[![Version](https://img.shields.io/npm/v/@waiting/base64.svg)](https://www.npmjs.com/package/@waiting/base64)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://opensource.org/licenses/MIT)
![](https://img.shields.io/badge/lang-TypeScript-blue.svg)
[![Build Status](https://travis-ci.org/waitingsong/base64.svg?branch=master)](https://travis-ci.org/waitingsong/base64)
[![Build status](https://ci.appveyor.com/api/projects/status/wp4a72sj7bc5ao3t/branch/master?svg=true)](https://ci.appveyor.com/project/waitingsong/base64/branch/master)
[![Coverage Status](https://coveralls.io/repos/github/waitingsong/base64/badge.svg?branch=master)](https://coveralls.io/github/waitingsong/base64?branch=master)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)


## Features
- Supports typeof `string`, `number` and [`bigint`](https://github.com/tc39/proposal-bigint)
- Supports `ArrayBuffer` or `Uint8Array`
- Encoding/Decoding via `TextEncoder`/`TextDecoder` under browser and `Buffer` under Node.js
- Building various modules`ESM`, `UMD` and `CJS`


## Browser requirement
- Ability of [TextEncoder](https://developer.mozilla.org/en-US/docs/Web/API/TextEncoder#Browser_compatibility) and 
  [TextDecoder](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder#Browser_compatibility)
- Polyfill [text-encoding](https://github.com/inexorabletash/text-encoding) for Edge


## Node.js requirement
- v10.4.0+ for `bigint` base64 encoding
- [ICU](https://nodejs.org/api/util.html#util_whatwg_supported_encodings) installation for base64 decoding with char encodings other then `utf-8` and `utf-16le`


## Installing
```bash
npm install @waiting/base64
```


## Usage of Node.js

See the [Docs](https://waitingsong.github.io/base64/) for details

### Encoding
```ts
import { b64encode, b64fromBuffer, b64urlEncode } from '@waiting/base64'

b64encode('A') === 'QQ=='
b64encode('schöne') === 'c2Now7ZuZQ=='
b64encode(1n) === b64encode(1) // bigint -> 'MQ==' 
b64encode('𠮷') === b64encode('\uD842\uDFB7') === b64encode('\u{20BB7}') // '8KCutw=='

const u8arr = Uint8Array.from([0xe4, 0xb8, 0xad, 0xe6, 0x96, 0x87])
b64fromBuffer(u8arr) === b64encode('中文')  // '5Lit5paH'

// URL-safe
b64urlEncode('A') === 'QQ'
b64urlEncode('中文测试') === '5Lit5paH5rWL6K-V'
```

### Decoding
```ts
import { b64decode, b64urlDecode } from '@waiting/base64'

b64decode('MQ==') === '1'
b64urlDecode('MQ') === '1'
```

### Transfer
```ts
import { b64toURLSafe, b64fromURLSafe } from '@waiting/base64'

// base64 -> URL-safe base64
b64toURLSafe('QQ==') === 'QQ'
b64toURLSafe('5Lit5paH5rWL6K+V') === '5Lit5paH5rWL6K-V'

// URL-safe base64 -> base64
b64fromURLSafe('QQ') === 'QQ=='
b64fromURLSafe('0J_RgNC40LLQtdGCLCDQvNC40YAh') === '0J/RgNC40LLQtdGCLCDQvNC40YAh'
```


## Usage of browser

See the [Docs](https://waitingsong.github.io/base64/) for details

### ESM
```html
<script type="module">
  import { b64encode, b64urlEncode } from './base64.esm.min.js' 
  
  console.log( b64encode('A') ) // 'QQ=='
  console.log( b64urlEncode('A') )  // 'QQ'
  console.log( b64encode('\uD842\uDFB7') ) // '8KCutw=='
</script>
```

### UMD
```html
<!-- polyfill for Edge
  Note: text-encoder-lite parse 4-bytes UTF-8 char incorrectly,
  so use text-encoding instead.
  see: https://github.com/solderjs/TextEncoderLite/issues/16
-->
<!--
<script src="https://raw.githubusercontent.com/inexorabletash/text-encoding/master/lib/encoding-indexes.js"></script>
<script src="https://raw.githubusercontent.com/inexorabletash/text-encoding/master/lib/encoding.js"></script>
-->
<script src="./base64.umd.min.js"></script>
<script>
  // global variable base64
  console.log( base64.b64encode('A') )
  console.log( base64.b64urlEncode('A') )
  console.log( base64.b64encode('\uD842\uDFB7') )
</script>
```


## Testing
```sh
npm run test

# for browsers FireFox and Chrome
npm run test:browser

# for browsers all available on this client
cd .config && npm i
npm run browser:detect
```


## Demo
- [Browser](https://github.com/waitingsong/base64/blob/master/test_browser/)
- [Node.js](https://github.com/waitingsong/base64/blob/master/test/)


## License
[MIT](LICENSE)


### Languages
- [English](README.md)
- [中文](README.zh-CN.md)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [4.2.9](https://github.com/waitingsong/base64/compare/v4.2.8...v4.2.9) (2019-07-15)


### Tests

* for '\u{10121}' ([fb4ab93](https://github.com/waitingsong/base64/commit/fb4ab93))



### [4.2.8](https://github.com/waitingsong/base64/compare/v4.2.6...v4.2.8) (2019-06-15)


### Tests

* update log ([ebffed1](https://github.com/waitingsong/base64/commit/ebffed1))
* **browser:** add 30_helper.test.ts ([13a8844](https://github.com/waitingsong/base64/commit/13a8844))
* **browser:** add 40_browser-textencoder-polyfill.test.ts, and skipped ([6e2eb21](https://github.com/waitingsong/base64/commit/6e2eb21))



### [4.2.6](https://github.com/waitingsong/base64/compare/v4.2.5...v4.2.6) (2019-06-14)


### Tests

* update cased of parseTextEncoder(), parseTextDecoder() ([1e2d625](https://github.com/waitingsong/base64/commit/1e2d625))



### [4.2.5](https://github.com/waitingsong/base64/compare/v4.2.4...v4.2.5) (2019-06-14)


### Tests

* validateDecoder() ([2f10583](https://github.com/waitingsong/base64/commit/2f10583))



### [4.2.4](https://github.com/waitingsong/base64/compare/v4.2.3...v4.2.4) (2019-06-13)



### [4.2.3](https://github.com/waitingsong/base64/compare/v4.2.2...v4.2.3) (2019-06-13)



### [4.2.2](https://github.com/waitingsong/base64/compare/v4.2.1...v4.2.2) (2019-06-13)



### [4.2.1](https://github.com/waitingsong/base64/compare/v4.2.0...v4.2.1) (2019-06-13)


### Build System

* use base64 as name of package for building ESM ([db96fe8](https://github.com/waitingsong/base64/commit/db96fe8))


### Tests

* **browser:** polyfill for ie11 ([5976675](https://github.com/waitingsong/base64/commit/5976675))



## [4.2.0](https://github.com/waitingsong/base64/compare/v4.1.0...v4.2.0) (2019-06-13)


### Features

* b64urlFromBuffer() ([0652596](https://github.com/waitingsong/base64/commit/0652596))



## [4.1.0](https://github.com/waitingsong/base64/compare/v4.0.1...v4.1.0) (2019-06-12)


### Features

* b64fromURLSafe() Convert URL-safe base64 string to base64 string ([cf267ff](https://github.com/waitingsong/base64/commit/cf267ff))
* b64urlDecode() Decode URL-safe base64 to original string ([a30ba9d](https://github.com/waitingsong/base64/commit/a30ba9d))
* b64urlEncode() Encode to URL-safe base64 ([0129fc7](https://github.com/waitingsong/base64/commit/0129fc7))



### [4.0.1](https://github.com/waitingsong/base64/compare/v4.0.0...v4.0.1) (2019-06-12)



## [4.0.0](https://github.com/waitingsong/base64/compare/v3.0.1...v4.0.0) (2019-06-12)


### Bug Fixes

* start should less then end for encodeChunk() ([1da0769](https://github.com/waitingsong/base64/commit/1da0769))


### Features

* b64toURLSafe() convert base64 string to URL-safe string ([fdd0300](https://github.com/waitingsong/base64/commit/fdd0300))
* export testB64(), testB64URL, validB64URLChars() ([5863162](https://github.com/waitingsong/base64/commit/5863162))


### Tests

* for validateB64(), validateB64URL() ([65306d3](https://github.com/waitingsong/base64/commit/65306d3))



### [3.0.1](https://github.com/waitingsong/base64/compare/v3.0.0...v3.0.1) (2019-06-12)



## [3.0.0](https://github.com/waitingsong/base64/compare/v2.2.0...v3.0.0) (2019-06-12)


### Bug Fixes

* error message with getLens() ([7ca94fb](https://github.com/waitingsong/base64/commit/7ca94fb))


### Features

* zero length of base64 not allowed of getLens() ([52e0be1](https://github.com/waitingsong/base64/commit/52e0be1))


### Tests

* fix import name ([a79854f](https://github.com/waitingsong/base64/commit/a79854f))



## [2.2.0](https://github.com/waitingsong/base64/compare/v2.1.0...v2.2.0) (2019-06-11)


### Bug Fixes

* wrong validating function within parseTextDecoder() ([b7db14e](https://github.com/waitingsong/base64/commit/b7db14e))


### Features

* detect ArrayBuffer instance via isArrayBuffer() ([930139a](https://github.com/waitingsong/base64/commit/930139a))
* detect Uint8Array instance via isUint8Array() ([164c6cf](https://github.com/waitingsong/base64/commit/164c6cf))
* validate input value for fromBuffer() ([c5f8a72](https://github.com/waitingsong/base64/commit/c5f8a72))
* validate input value for fromBuffer() ([4ce5c24](https://github.com/waitingsong/base64/commit/4ce5c24))


### Tests

* add 30_helper.test.ts ([9d7ac1b](https://github.com/waitingsong/base64/commit/9d7ac1b))
* add case for encodeChunk() ([f90d3e8](https://github.com/waitingsong/base64/commit/f90d3e8))
* fix import module ([686eb11](https://github.com/waitingsong/base64/commit/686eb11))
* update cases of b64decode() ([8f96c13](https://github.com/waitingsong/base64/commit/8f96c13))



## [2.1.0](https://github.com/waitingsong/base64/compare/v2.0.0...v2.1.0) (2019-06-11)


### Features

* b64byteLength() ([477fe7b](https://github.com/waitingsong/base64/commit/477fe7b))



## [2.0.0](https://github.com/waitingsong/base64/compare/v1.1.0...v2.0.0) (2019-06-11)



## [1.1.0](https://github.com/waitingsong/base64/compare/v1.0.0...v1.1.0) (2019-06-11)


### Features

* b64FromBuffer() encoding from ArrayBuffer or Uint8Array input ([236feda](https://github.com/waitingsong/base64/commit/236feda))



## 1.0.0 (2019-06-11)


### Features

* b64encode() and b64decode() ([09f44c1](https://github.com/waitingsong/base64/commit/09f44c1))


test_browser/config.ts
import {
  baseChars,
} from '../src/lib/config'

import { _dummy } from './patch'


// @ts-ignore
export const isIE = /* @cc_on!@*/false || !! document.documentMode
// @ts-ignore
export const isEdge = ! isIE && !! window.StyleMedia

export const input1: Array<string | number | bigint> = [
  baseChars,
  'a\'b"c<d>e&f=g/[].,:;+-_`~!@#$%^&*(){}*\\ ',
  'A',
  '  ',
  ' ',
  '\r\n',
  '\r',
  '\n',
  '0',
  '€',
  'schöne',
  'grüße',
  '中文，。！',
  '联通',
  'Привет, мир!',
  Number.MAX_SAFE_INTEGER,
  Number.MIN_SAFE_INTEGER,
  1,
  -1,
  0,
  -0,
]
export const base1: string[] = [
  'QUJDREVGR0hJSktMTU5PUFFSU1RVVldYWVphYmNkZWZnaGlqa2xtbm9wcXJzdHV2d3h5ejAxMjM0NTY3ODkrLw==',
  'YSdiImM8ZD5lJmY9Zy9bXS4sOjsrLV9gfiFAIyQlXiYqKCl7fSpcIA==',
  'QQ==',
  'ICA=',
  'IA==',
  'DQo=',
  'DQ==',
  'Cg==',
  'MA==',
  '4oKs',
  'c2Now7ZuZQ==',
  'Z3LDvMOfZQ==',
  '5Lit5paH77yM44CC77yB',
  '6IGU6YCa',
  '0J/RgNC40LLQtdGCLCDQvNC40YAh',
  'OTAwNzE5OTI1NDc0MDk5MQ==',
  'LTkwMDcxOTkyNTQ3NDA5OTE=',
  'MQ==',
  'LTE=',
  'MA==',
  'MA==',
]

export const input2: string[] = [
  '𠮷', '\uD842\uDFB7', '\u{D842}\u{DFB7}', '\u{20BB7}',
  '\u{10121}',
]
export const base2: string[] = [
  '8KCutw==',
  '8KCutw==',
  '8KCutw==',
  '8KCutw==',
  '8JCEoQ==',
]

export const input3: string[] = [
  '\uD800',
  '\uD801',
  '\uD842',
]
export const base3: string[] = [
  '77+9',
  '77+9',
  '77+9',
]

export const input4: Array<string | number> = [
  Math.random(),
  Math.random().toString(),
]
export const base4: string[] = [
  'MC41NDc0MjU2NzQ0NTY2NzA4',
  'MC4xMzk2OTgxNzY3NTE1Njg4MQ==',
]


export const input44: any[] = [
  {},
  null,
  true,
  false,
  (void 0),
]

export type Input8Item = [number[], string]
export const input8: Input8Item[] = [
  [ [0, 0, 0], 'AAAA'],
  [ [0, 0, 1], 'AAAB'],
  [ [0, 1, -1], 'AAH/'],
  [ [1, 1, 1], 'AQEB'],
  [ [0, -73, 23], 'ALcX'],
  [ [0xe4, 0xb8, 0xad, 0xe6, 0x96, 0x87], '5Lit5paH'],
]

export const input9 = [
  ' ',
  '*',
  '?',
]


export const inputTypedArrayExcludingUint8Array = [
  new Uint16Array(),
  new Uint32Array(),
  new Uint8ClampedArray(),
  new Int8Array(),
  new Int16Array(),
  new Int32Array(),
  new Float32Array(),
  // new Float64Array(),
  // new BigInt64Array(),
]

export const inputUint8Array = [
  new Uint8Array(0),
  new Uint8Array(1),
  new Uint8Array(1024),
]

export const inputArrayBuffer = [
  new ArrayBuffer(0),
  new ArrayBuffer(1),
  new ArrayBuffer(1024),
]


/** [string, base64, URL-safe base64] */
export const inputURLSafe = [
  ['中文测试', '5Lit5paH5rWL6K+V', '5Lit5paH5rWL6K-V'],
  ['Привет, мир!', '0J/RgNC40LLQtdGCLCDQvNC40YAh', '0J_RgNC40LLQtdGCLCDQvNC40YAh'],
  ['A', 'QQ==', 'QQ'],
  ['12345678901234567890', 'MTIzNDU2Nzg5MDEyMzQ1Njc4OTA=', 'MTIzNDU2Nzg5MDEyMzQ1Njc4OTA'],
]

export const inputURLSafe2 = [
  'A',
  'QQ',
  '=QQ',
  'Q=Q',
  'AABB\nAABB',
]

export const inputBase64CharsInvalid = [
  '',
  '!@(^*',
  'AB!!',
  'QQ===',
  '=QQ==',
  '=QQ=',
  '=QQ',
  'AABB\nAABB',
]
export const inputBase64Invalid = [
  0,
  '',
  'Q',
  'QQ',
  'QQQ',
  'QQQQQ',
  'Q=QQ',
  'QQQ==',
  'QQ===',
  '=QQ==',
  '=QQ=',
  '=QQ',
  'AABB\nAABB',
]

export const inputBase64URLCharsInvalid = [
  '',
  '!@(^*',
  'AB!!',
  'QQ===',
  'QQ==',
  '=QQ==',
  '=QQ',
  'QQ+/',
  'AABB\nAABB',
]
export const inputBase64URLInvalid = [
  0,
  '',
  'Q',
  'AABB\nAABB',
]

export const inputURLSafe3: Array<[number[], string]> = [
  [ [0xff, 0xff, 0xbe, 0xff, 0xef, 0xbf, 0xfb, 0xef, 0xff], '__--_--_--__'],
]


export const inputInvalidEncoderAndDecoder: any[] = [
  1, 'a', false, true, null,
  {},
  (void 0),
]
